### ISSUE TITLE
Bloons TD 5 - rounds do not end correctly, and there is sometimes no starting tower unlocked

### ISSUE THREAD

**Issue by arthomnix** [ruffle-rs/ruffle#14947]
### Describe the bug

In Bloons TD 5, rounds do not end once all bloons have been popped. Instead, this error message occurs when the round is supposed to end:

```
2024-01-25T19:18:20.296887Z ERROR ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "TIMEEVENT_COMPLETE", class: events::TimerEvent, ptr: 0x62105b20b0f0 }) to handler FunctionObject(FunctionObject { ptr: 0x62105b397450, name: Ok("MethodInfo-513()") }) : TypeError: Error #1010: A term is undefined and has no properties. (accessing field: flavorText)
```
![20240125_19h19m24s_grim](https://github.com/ruffle-rs/ruffle/assets/35371030/0164b37f-02bd-4333-9b30-d35f4cf02249)

Additionally, there is a chance that there will be no towers unlocked at the start of the game, making it impossible to play.

Correct behaviour (dart monkey unlocked at start):
![20240125_19h17m08s_grim](https://github.com/ruffle-rs/ruffle/assets/35371030/4fecc08d-a028-4ee3-9f45-d35621447ccf)

Incorrect behaviour (no towers unlocked at start):
![20240125_19h23m29s_grim](https://github.com/ruffle-rs/ruffle/assets/35371030/92d6db83-84da-48e2-a42c-fbe28e7fd9b7)

### Steps to reproduce

1. Open BTD5 and start a new game in any level (there is a chance that there will be no towers unlocked)
2. Start the first round by pressing Go
3. The round will never finish, and the message above will be printed to the log

(There is a small chance that the round does actually finish correctly (it happened once during my testing), if this happens just start another round)

### Expected behavior

Every round ends after all bloons have been popped or gone all the way through the stage (the button changes back to Go), and the dart monkey is always unlocked at the start of the game.

### Content Location

[btd5.swf.gz](https://github.com/ruffle-rs/ruffle/files/14057097/btd5.swf.gz)

This is the latest version directly from the NinjaKiwi website, that I modified slightly to remove the site restrictions. It does work fine using the official desktop Flash Player. I'm not sure if this issue occurs on other versions of the game - from my limited testing, the version at https://data.gameflare.com/games/3550/WDzRSvadTXBNOR.swf seems not to have this issue.

### Affected platform

Desktop app

### Operating system

Arch Linux (kernel 6.7.1-zen1-1-zen, Hyprland 0.34.0)

**Comment by arthomnix** [comment]
This bug also appears on [the original unmodified SWF](https://web.archive.org/web/20190327172214/http://assets.ninjakiwi.com/Games/gameswfs/btd5-dat.swf?v=20190114), but you have to pass `--spoof-url "http://assets.ninjakiwi.com/Games/gameswfs/btd5-dat.swf?v=20190114"` for it to let you start a game at all

**Comment by sombraguerrero** [comment]
I will not pretend to know this game's codebase well enough to actually tackle the underlying issue, however, through decompilation and a very long play session, I have confirmed that the easiest way to circumvent this bug is to simply not have in-game tips enabled. There is an if statement on line 1040 of gui.ingame.InGameMenu that precedes an assignment of the variable _loc3_. This assignment is clearly failing, but one of the conditions of the if statement includes whether or not you have tips enabled when you're not in challenge mode. If this assignment is skipped, the normal game mode will proceed through rounds unhindered.

**Comment by sombraguerrero** [comment]
Seems to be a discrepancy between Flash's implementation of Math.random and ours. The array index involved in the aforementioned assignment is returning as an overflow/underflow value.

**Comment by arthomnix** [comment]
Should be fixed by #14982

**Comment by sombraguerrero** [comment]
Been testing with Aaron's branch and unsurprisingly, this seems to fix several additional things that were less critically, but nevertheless wrong with the gameplay.
